File: tests/src/test/kotlin/io/github/expo/kolibri/tests/RegistrationTests.kt
package io.github.expo.kolibri.tests

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Every registration shape native_method.h supports: free functions (auto and template-signature),
 * captureless lambdas (auto and explicit signature), receiver plumbing (leading `JNIEnv*`,
 * `JavaReceiver`, both, neither), and `Ref`/`GlobalRef` returns lowered to Java objects.
 * Member-function bindings are covered by NativeObjectTests. The natives live in
 * registration_tests.cpp and are registered through NativeMethodsBuilder like every other suite.
 */
class RegistrationTests : NativeTestBase() {
  private external fun freeFunctionEcho(value: Int): Int
  private external fun freeFunctionWithSignatureTemplate(value: Long): Long
  private external fun lambdaConcat(a: String, b: String): String
  private external fun lambdaWithExplicitSignature(value: Int): Int
  private external fun envOnly(): Boolean
  private external fun receiverOnly(): Boolean
  private external fun envAndReceiver(): Boolean
  private external fun noEnvNoReceiver(a: Int, b: Int): Int
  private external fun refReturn(): String
  private external fun nullableRefReturn(returnNull: Boolean): String?
  private external fun globalRefReturn(): String

  @Test
  fun `a free function binds with an auto-generated signature`() {
    assertEquals(7, freeFunctionEcho(7))
  }

  @Test
  fun `a free function binds with a signature supplied as a template argument`() {
    assertEquals(1L shl 40, freeFunctionWithSignatureTemplate(1L shl 40))
  }

  @Test
  fun `a captureless lambda binds with an auto-generated signature`() {
    assertEquals("foobar", lambdaConcat("foo", "bar"))
  }

  @Test
  fun `a captureless lambda binds with an explicit signature string`() {
    assertEquals(42, lambdaWithExplicitSignature(41))
  }

  @Test
  fun `a leading JNIEnv parameter is injected, not read from Java arguments`() {
    assertTrue(envOnly())
  }

  @Test
  fun `a JavaReceiver parameter receives the Java receiver`() {
    assertTrue(receiverOnly())
  }

  @Test
  fun `JNIEnv and JavaReceiver can both be requested`() {
    assertTrue(envAndReceiver())
  }

  @Test
  fun `a binding with neither env nor receiver sees only the Java arguments`() {
    assertEquals(5, noEnvNoReceiver(2, 3))
  }

  @Test
  fun `a returned local Ref is released to Java as the result object`() {
    assertEquals("owned local", refReturn())
  }

  @Test
  fun `an empty returned Ref becomes null`() {
    assertNull(nullableRefReturn(true))
    assertEquals("present", nullableRefReturn(false))
  }

  @Test
  fun `a returned GlobalRef is lowered to a fresh local reference`() {
    assertEquals("from global", globalRefReturn())
  }
}

File: tests/src/test/cpp/registration_tests.cpp
#include <string>

#include <kolibri/JavaReceiver.h>
#include <kolibri/Ref.h>
#include <kolibri/native_method.h>
#include <kolibri/string_utils.h>

#include "expect.h"
#include "test_helpers.h"

namespace expo::kolibri::tests {
  namespace {
    jint freeEcho(jint value) noexcept { return value; }

    jlong freeEchoLong(jlong value) noexcept { return value; }
  } // namespace

  void registerRegistrationTests(JNIEnv* env) {
    registerNative(env, "io/github/expo/kolibri/tests/RegistrationTests")
        .method<&freeEcho>("freeFunctionEcho")
        .method<&freeEchoLong, jlong(jlong)>("freeFunctionWithSignatureTemplate")
        .method(
          "lambdaConcat",
          [](JNIEnv* e, jstring a, jstring b) -> jstring {
            return toJString(e, toStdString(e, a) + toStdString(e, b));
          }
        )
        .method("lambdaWithExplicitSignature", "(I)I", [](jint value) -> jint { return value + 1; })
        .method("envOnly", [](JNIEnv* e) -> jboolean { return e != nullptr ? JNI_TRUE : JNI_FALSE; })
        .method(
          "receiverOnly",
          [](JavaReceiver self) -> jboolean { return self.self != nullptr ? JNI_TRUE : JNI_FALSE; }
        )
        .method(
          "envAndReceiver",
          [](JNIEnv* e, JavaReceiver self) -> jboolean {
            KOLIBRI_EXPECT(e != nullptr);
            KOLIBRI_EXPECT(self.self != nullptr);
            return JNI_TRUE;
          }
        )
        .method("noEnvNoReceiver", [](jint a, jint b) noexcept -> jint { return a + b; })
        .method(
          "refReturn",
          [](JNIEnv* e) -> Ref<jstring> {
            return Ref<jstring>::adopt(e, toJString(e, std::string("owned local")));
          }
        )
        .method(
          "nullableRefReturn",
          [](JNIEnv* e, jboolean returnNull) -> Ref<jstring> {
            if (returnNull == JNI_TRUE) {
              return {};
            }
            return Ref<jstring>::adopt(e, toJString(e, std::string("present")));
          }
        )
        .method(
          "globalRefReturn",
          [](JNIEnv* e) -> GlobalRef<jstring> {
            const auto local =
                Ref<jstring>::adopt(e, toJString(e, std::string("from global")));
            return GlobalRef<jstring>::make(e, local.get());
          }
        )
        .commit();
  }
} // namespace expo::kolibri::tests
